refactor: Copy sequence lengths once in decoder setup

Funatiq opened the pull request:
## Description

* Copy sequence lengths once in `setupDecoderStep` instead of in each `decoderStepAsync` iteration.
* Refactor `GenerateRequestOptions` using more fine-grained functions.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
refactor: Copy sequence lengths once in decoder setup

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #4250](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4250/) [ run ] triggered by Bot

Funatiq committed:
refactor: Move getEmbeddingBias to anonymous namespace

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Filter context requests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Update DecoderInputBuffers to remove duplicated buffers

- Renamed and reorganized buffer variables in decoderBuffers.h and decoderBuffers.cpp for better readability.
- Adjusted references in generateRequestOptions.cpp to align with the new buffer structure.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: GenerateRequestOptions using more fine-grained functions

- Added a new method `createDecoderRequests` to encapsulate the logic for creating decoder requests from finished context requests.
- Updated the `operator()` method to utilize the new method, improving code clarity and maintainability.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #4250](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4250/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #3038](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3038/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #4256](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4256/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4256](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4256/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3043](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3043/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --stage-list "A10-CPP-1, A30-CPP-1, A30-CPP-2, H100_PCIe-CPP-1"

tensorrt-cicd commented:
[PR_Github #4344](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4344/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4344](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4344/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3114 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3114/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #4390](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4390/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4388](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4388/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #4390](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4390/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `6b95307`

Funatiq committed:
refactor: Update TRTLLMDecoder

- Updated the `generate_request_options` call.
- Updated the `make_decoding_batch_input_output` call.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #4391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4391/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4391](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4391/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3154](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3154/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #4866](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4866/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3524](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3524/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run --disable-fail-fast

Funatiq committed:
refactor: Remove const where we modify input buffers

- Changed `DecoderInputBuffers` parameters from const references to non-const references in multiple functions to allow modifications.
- Updated related function calls to ensure compatibility with the new parameter types.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented on cpp/tensorrt_llm/batch_manager/generateRequestOptions.cpp:
```diff
@@ -16,67 +16,163 @@
+namespace
+{
+
+void copySequenceLengths(RequestVector const& contextRequests, DecoderInputBuffers const& inputBuffers,
```
Done.

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #5335](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5335/) [ run ] triggered by Bot

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #5355](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5355/) [ run ] triggered by Bot

Funatiq committed:
fixup! refactor: Copy sequence lengths once in decoder setup

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #5361](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5361/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5361](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5361/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3914](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3914/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #5473](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5473/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #3996](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3996/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #5514](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5514/) [ skip ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5514](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5514/) [ skip ] completed with state `SUCCESS`
Skipping testing for commit `2cf5667`
